src/client/pkg/grpcutil/server.go

```diff
@@ -44,62 +47,90 @@ type ServerOptions struct {
-			grpc.KeepaliveEnforcementPolicy(keepalive.EnforcementPolicy{
-				MinTime:             5 * time.Second,
-				PermitWithoutStream: true,
-			}),
-			grpc.UnaryInterceptor(tracing.UnaryServerInterceptor()),
-			grpc.StreamInterceptor(tracing.StreamServerInterceptor()),
-		}
-		if server.PublicPortTLSAllowed {
-			// Validate environment
-			certPath, keyPath, err := tls.GetCertPaths()
-			if err != nil {
-				log.Warnf("TLS disabled: %v", err)
-			} else {
-				// Read TLS cert and key
-				transportCreds, err := credentials.NewServerTLSFromFile(certPath, keyPath)
+		eg.Go(func() error {
+			if server.RegisterFunc == nil {
+				return ErrMustSpecifyRegisterFunc
+			}
+			if server.Port == 0 && !server.AnyPort {
+				return ErrMustSpecifyPort
+			}
+			opts := []grpc.ServerOption{
+				grpc.MaxConcurrentStreams(math.MaxUint32),
+				grpc.MaxRecvMsgSize(server.MaxMsgSize),
+				grpc.MaxSendMsgSize(server.MaxMsgSize),
+				grpc.KeepaliveEnforcementPolicy(keepalive.EnforcementPolicy{
+					MinTime:             5 * time.Second,
+					PermitWithoutStream: true,
+				}),
+				grpc.UnaryInterceptor(tracing.UnaryServerInterceptor()),
+				grpc.StreamInterceptor(tracing.StreamServerInterceptor()),
+			}
+			if server.PublicPortTLSAllowed {
+				// Validate environment
+				certPath, keyPath, err := tls.GetCertPaths()
 				if err != nil {
-					return fmt.Errorf("couldn't build transport creds: %v", err)
+					log.Warnf("TLS disabled: %v", err)
+				} else {
+					// Read TLS cert and key
+					transportCreds, err := credentials.NewServerTLSFromFile(certPath, keyPath)
+					if err != nil {
+						return fmt.Errorf("couldn't build transport creds: %v", err)
+					}
+					opts = append(opts, grpc.Creds(transportCreds))
 				}
-				opts = append(opts, grpc.Creds(transportCreds))
 			}
-		}
-
-		grpcServer := grpc.NewServer(opts...)
-		if err := server.RegisterFunc(grpcServer); err != nil {
-			return err
-		}
-		listener, err := net.Listen("tcp", fmt.Sprintf(":%d", server.Port))
-		if err != nil {
-			return err
-		}
-		if server.Cancel != nil {
-			go func() {
-				<-server.Cancel
-				if err := listener.Close(); err != nil {
-					fmt.Printf("listener.Close(): %v\n", err)
-				}
-			}()
-		}
-		if err := grpcServer.Serve(listener); err != nil {
-			return err
-		}
+
+			grpcServer := grpc.NewServer(opts...)
+			if err := server.RegisterFunc(grpcServer); err != nil {
+				return err
+			}
+			port := server.Port
+			if server.AnyPort {
+				port = 0
+			}
+			listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", server.Host, port))
+			if err != nil {
+				return err
+			}
+
+			eg.Go(func() error {
+				<-ctx.Done()
+				grpcServer.GracefulStop() // This also closes the TCP listener
+				return nil
+			})
+
+			serverRuns = append(serverRuns, &ServerRun{
+				GrpcServer: grpcServer,
+				Listener:   listener,
+			})
+
+			ready <- true
+
+			return grpcServer.Serve(listener)
+		})
 	}
-	return nil
+
+	for range servers {
+		<-ready
```

Comment:
I think this needs to be a select on `ready` and `ctx.Done()` so in case the ctx gets cancelled during the startup process it doesn't deadlock. Not 100% sure, because it looks like the `ctx` might not actually be used until after the ready signal is sent. I'd add it anyway though just to be defensive, I could see this code changing and adding something that touches the context.